### PR TITLE
Refactor recording core, mock tool instead of mock invoke tool

### PR DESCRIPTION
**Pull request by crazygao**
# Description

Refactor recording core, mock tool instead of mock invoke tool.
recording could yield two field.
it could return a recording status.
you could customize recording_array, and set which tool could be recorded in the test.

# All Promptflow Contribution checklist:
- [x] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow/tests/conftest.py [hunk 1]
```diff
@@ -8,14 +8,14 @@
 
 import pytest
 from _constants import (
+    CLI_PERF_MONITOR_AGENT,
```
Comment (zhengfeiwang) [inline src/promptflow/tests/conftest.py:11]: I understand this change is done by pre-commit? if so, maybe we still miss some steps in dev setup...

File: src/promptflow/tests/sdk_cli_test/e2etests/test_cli.py [hunk 2]
```diff
@@ -1421,7 +1421,7 @@ def test_tool_list(self, capsys):
         run_pf_command("tool", "list")
         outerr = capsys.readouterr()
         tools_dict = json.loads(outerr.out)
-        package_tool_name = "promptflow.tools.embedding.embedding"
+        package_tool_name = "promptflow.tools.open_source_llm.OpenSourceLLM.call"
```
Comment (zhengfeiwang) [inline src/promptflow/tests/sdk_cli_test/e2etests/test_cli.py:1424]: why update this?

File: src/promptflow/tests/sdk_cli_test/recording_utilities/mock_tool.py [hunk 3]
```diff
@@ -0,0 +1,194 @@
+import functools
```
Comment (zhengfeiwang) [inline src/promptflow/tests/sdk_cli_test/recording_utilities/mock_tool.py:1]: maybe add some comments for these functions, if you will do an update

File: src/promptflow/tests/conftest.py (hunk 1 above)
Comment (crazygao) [reply src/promptflow/tests/conftest.py:11]: Yep

File: src/promptflow/tests/sdk_cli_test/recording_utilities/mock_tool.py (hunk 3 above)
Comment (crazygao) [reply src/promptflow/tests/sdk_cli_test/recording_utilities/mock_tool.py:1]: OKay

File: src/promptflow/tests/sdk_cli_test/e2etests/test_cli.py (hunk 2 above)
Comment (crazygao) [reply src/promptflow/tests/sdk_cli_test/e2etests/test_cli.py:1424]: let me double check

**Review by zhengfeiwang — APPROVED** [review]

File: src/promptflow/tests/sdk_cli_test/conftest.py [hunk 4]
```diff
@@ -202,10 +153,16 @@ def recording_file_override(request: pytest.FixtureRequest, mocker: MockerFixtur
...
+
     if RecordStorage.is_replaying_mode() or RecordStorage.is_recording_mode():
-        original_fun = FlowExecutionContext._invoke_tool_with_timer
-        mocker.patch(
-            "promptflow._core.flow_execution_context.FlowExecutionContext._invoke_tool_with_timer",
-            mock_origin(original_fun),
-        )
-    yield
+        mocked_tool = mock_tool(original_tool)
+        mocker.patch("promptflow._core.tool.tool", mocked_tool)
```
Comment (wangchao1230) [inline src/promptflow/tests/sdk_cli_test/conftest.py:160]: so strange we have three places to patch...

**Review by wangchao1230 — APPROVED** [review]

**Pull request merged**